test/spec/modules/conversantAnalyticsAdapter_spec.js

```diff
@@ -0,0 +1,733 @@
+      expect(cacheObj.bar).to.be.undefined;
+    });
+  });
+
+  describe('Bid Timeout Event Tests', function() {
+    const bidTimeoutPayload = [{
+      'bidId': '80882409358b8a8',
+      'bidder': 'conversant',
+      'adUnitCode': 'MedRect',
+      'auctionId': 'afbd6e0b-e45b-46ab-87bf-c0bac0cb8881'
+    }, {
+      'bidId': '9da4c107a6f24c8',
+      'bidder': 'conversant',
+      'adUnitCode': 'Leaderboard',
+      'auctionId': 'afbd6e0b-e45b-46ab-87bf-c0bac0cb8881'
+    }
+    ];
+
+    beforeEach(function () {
+      conversantAnalytics.enableAnalytics(VALID_ALWAYS_SAMPLE_CONFIG);
+    });
+
+    afterEach(function () {
+      conversantAnalytics.disableAnalytics();
+    });
+
+    it('should put both items in timeout cache', function() {
+      expect(Object.keys(caObj.timeoutCache).length).to.equal(0);
+      events.emit(constants.EVENTS.BID_TIMEOUT, bidTimeoutPayload);
+      expect(Object.keys(caObj.timeoutCache).length).to.equal(2);
+      expect(requests.length).to.equal(0);
+    });
+  });
+
+  describe('Render Failed Tests', function() {
+    const RENDER_FAILED_PAYLOAD = {
+      reason: 'reason',
+      message: 'value',
+      adId: '57e03aeafd83a68'
+    };
+
+    const RENDER_FAILED_PAYLOAD_NO_ADID = {
+      reason: 'reason',
+      message: 'value'
+    };
+
+    beforeEach(function () {
+      conversantAnalytics.enableAnalytics(VALID_ALWAYS_SAMPLE_CONFIG);
+    });
+
+    afterEach(function () {
+      conversantAnalytics.disableAnalytics();
+    });
+
+    it('should empty adIdLookup and send data', function() {
+      caObj.adIdLookup[RENDER_FAILED_PAYLOAD.adId] = {
+        bidderCode: 'bidderCode',
+        adUnitCode: 'adUnitCode',
+        auctionId: 'auctionId',
+        timeReceived: Date.now()
+      };
+
+      expect(Object.keys(caObj.adIdLookup).length).to.equal(1);
+      events.emit(constants.EVENTS.AD_RENDER_FAILED, RENDER_FAILED_PAYLOAD);
+      expect(Object.keys(caObj.adIdLookup).length).to.equal(0); // object should be removed
+      expect(requests.length).to.equal(1);
+      const data = JSON.parse(requests[0].requestBody);
+
+      expect(data.auction.auctionId).to.equal('auctionId');
+      expect(data.auction.preBidVersion).to.equal(PREBID_VERSION);
+      expect(data.auction.sid).to.equal(SITE_ID);
+      expect(data.adUnits['adUnitCode'].bids['bidderCode'].eventCodes.includes(caObj.CNVR_RENDER_FAILED)).to.be.true;
+      expect(data.adUnits['adUnitCode'].bids['bidderCode'].message).to.have.lengthOf.above(0);
+    });
+
+    it('should not send data if no adId', function() {
+      caObj.adIdLookup[RENDER_FAILED_PAYLOAD.adId] = {
+        bidderCode: 'bidderCode',
+        adUnitCode: 'adUnitCode',
+        auctionId: 'auctionId',
+        timeReceived: Date.now()
+      };
+
+      expect(Object.keys(caObj.adIdLookup).length).to.equal(1);
+      events.emit(constants.EVENTS.AD_RENDER_FAILED, RENDER_FAILED_PAYLOAD_NO_ADID);
+      expect(requests.length).to.equal(0);
```

Comment:
Verify request

Reply:
Just confirming that renderFailed event doesn't send data that doesn't match from the queue.

Reply:
Ah right.  The length checks in this one make sense.